### ISSUE TITLE
WinRM - only fails in the 1st run - Bad HTTP response returned from server. Code 500

### ISSUE THREAD

**Issue by msghaleb** [ansible/ansible#16873]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = /home/ubuntu/ansible-cloudfirst-master/plays/ansible.cfg
  configured module search path = ['../modules', '/usr/share/ansible']
```
##### CONFIGURATION

Nothing changed on the ansible.cfg file, this is a brand new testing ansible server.
##### OS / ENVIRONMENT

Ansible: Ubuntu 14.04 Latest update
Managing: Windows 2012 R2 / 2012
##### SUMMARY

I'm trying to provision a server using the EC2 module and then installing a software on it. It used to work until new days ago. Now when running the playfile it give a "Bad HTTP response returned from server. Code 500')" error JUST FOR THE 1ST TIME! if I rerun the same code again it works perfectly!! however if I destroy the server and spin it up again its the same behaviour. 

If you set gather facts to true, it will fail in the setup, and if you rerun it will work. (just rerun directly, no need to wait)

I tried on many different windows versions, and different AWS accounts. Constantly the same problem.
##### STEPS TO REPRODUCE

To reproduce just run the below play file, however you need to put in your ID, KEY, Security Group and AMI image ID. The error happens in the last task.

Here is my Inventory file:
[test]
ghaleb_server ansible_winrm_server_cert_validation=ignore

``` yml

---
- hosts: test
  gather_facts: False
  vars:
    aws_windows_user    : "Administrator"
    aws_windows_pass    : "SomePassW0rd!"
    aws_count_tag       :
           Name         : "TestingServer"
           Group        : "TestingServers"
    ansible_winrm_server_cert_validation : ignore
    ansible_winrm_transport : ssl
    ansible_ssh_port    : 5986
    ansible_ssh_user    : "{{ aws_windows_user }}"
    ansible_ssh_pass    : "{{ aws_windows_pass }}"
    ansible_user        : "{{ aws_windows_user }}"
    ansible_pass        : "{{ aws_windows_pass }}"
    test_user_data      : |
                            <powershell>
                            echo "Trying to set policy " > C:\bootscript.txt
                            Set-ExecutionPolicy RemoteSigned
                            echo "Downloading  ConfigureRemotingForAnsible.ps1" >> C:\bootscript.txt
                            $file = "C:\ConfigureRemotingForAnsible.ps1"
                            $webclient = New-Object System.Net.WebClient
                            $url = "https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1"
                            $webclient.DownloadFile($url,$file)
                            echo "Invoke-Expression  ConfigureRemotingForAnsible.ps1" >> C:\bootscript.txt
                            Invoke-Expression C:\ConfigureRemotingForAnsible.ps1
                            echo "Trying to change password for Admin" >> C:\bootscript.txt
                            net user {{ ansible_ssh_user }} {{ ansible_ssh_pass }}
                            echo "Bootstrap finished" >> C:\bootscript.txt
                            </powershell>
  tasks:
  - name: provision | creating a testing server in EC2
    connection: local
    ec2:
         aws_access_key: "YOUR_ACCESS_KEY"
         aws_secret_key: "YOUR_SEC_KEY"
         group_id: "sg-01a52568" # Change that one to a security group in your account/region
         instance_type: "t2.medium"
         image: "ami-09e61366" # you need to make sure you are using the correct ami file
         region: "eu-central-1"
         wait: true
         exact_count: 1
         count_tag: "{{ aws_count_tag | default(omit)}}"
         user_data: "{{ test_user_data }} "
         instance_tags: "{{ aws_count_tag | default(omit)}}"
    register: ec2

  - name: provision | Waiting for port "{{ ansible_ssh_port }}" to open
    connection: local
    wait_for:
        port="{{ ansible_ssh_port | default('5986') }}"
        host="{{ ec2.tagged_instances[0].public_ip }}"
        delay=1
        timeout=1200

  - name: provision | set facts for ansible_ssh_host
    connection: local
    set_fact:
        ansible_ssh_host: "{{ ec2.tagged_instances[0].public_ip }}"

  - name: provision | checking winrm connection
    connection: winrm
    vars:
      ansible_connection: winrm
    raw: whoami
```

##### EXPECTED RESULTS

Just run the code one more time, it will work fine! (very strange)

```
TASK [provision | checking winrm connection] ***********************************
task path: /home/ubuntu/ansible-cloudfirst-master/plays/test.yml:62
<52.59.81.233> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO 52.59.81.233
<52.59.81.233> WINRM CONNECT: transport=ssl endpoint=https://52.59.81.233:5986/wsman
<52.59.81.233> WINRM OPEN SHELL: 36CB7A73-0CA1-445B-8201-D5B9BE610554
<52.59.81.233> EXEC whoami
<52.59.81.233> WINRM EXEC 'PowerShell' ['-NoProfile', '-NonInteractive', '-ExecutionPolicy', 'Unrestricted', '-EncodedCommand', 'dwBoAG8AYQBtAGkA']
<52.59.81.233> WINRM RESULT u'<Response code 0, out "win-hk8d2cji282\\admi", err "">'
<52.59.81.233> WINRM STDOUT win-hk8d2cji282\administrator

<52.59.81.233> WINRM STDERR
<52.59.81.233> WINRM CLOSE SHELL: 36CB7A73-0CA1-445B-8201-D5B9BE610554
ok: [ghaleb_server] => {"changed": false, "invocation": {"module_args": {"_raw_params": "whoami"}, "module_name": "raw"}, "rc": 0, "stderr": "", "stdout": "win-hk8d2cji282\\administrator\r\n", "stdout_lines": ["win-hk8d2cji282\\administrator"]}
```
##### ACTUAL RESULTS

```
TASK [provision | checking winrm connection] ***********************************
task path: /home/ubuntu/ansible-cloudfirst-master/plays/test.yml:62
<52.59.81.233> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO 52.59.81.233
<52.59.81.233> WINRM CONNECT: transport=ssl endpoint=https://52.59.81.233:5986/wsman
<52.59.81.233> WINRM CONNECTION ERROR: (u'http', u'Bad HTTP response returned from server. Code 500')
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/connection/winrm.py", line 151, in _winrm_connect
    self.shell_id = protocol.open_shell(codepage=65001) # UTF-8
  File "/usr/local/lib/python2.7/dist-packages/winrm/protocol.py", line 132, in open_shell
    res = self.send_message(xmltodict.unparse(req))
  File "/usr/local/lib/python2.7/dist-packages/winrm/protocol.py", line 207, in send_message
    return self.transport.send_message(message)
  File "/usr/local/lib/python2.7/dist-packages/winrm/transport.py", line 191, in send_message
    raise WinRMTransportError('http', error_message)
WinRMTransportError: (u'http', u'Bad HTTP response returned from server. Code 500')

fatal: [ghaleb_server]: UNREACHABLE! => {"changed": false, "msg": "ssl: (u'http', u'Bad HTTP response returned from server. Code 500')", "unreachable": true}
```

**Comment by jhawkesworth** [comment]
Not tried this myself, but I wonder if the winrm service isn't ready (even though the port has become available) - have you tried inserting a pause for a while before running the raw: whoami ?

**Comment by msghaleb** [comment]
Yep, I've tried putting a 5 mins pause. I've also tried to put a loop (do until) for the whoiam task, no luck. I also tried waiting for 10 mins and then running the winrm tasks from a different file (Same result) it must fail for the first time with the error above.

**Comment by jhawkesworth** [comment]
That's really strange.
Another thing you could try is seeing if you get the same behavior over http instead of https (port 5985)?
I'd be tempted to add a 
ignore_errors: True 
onto the raw: whoami
call and see if that's enough to workaround this problem.
Out of interest, how long does it take to fail?  
Is there anything in the event log of the remote windows machine that records the 500 response.  I'd try looking for winrm or winrs event types, and also logon events as well.

**Comment by msghaleb** [comment]
Yes the remote machine shows a warning that the WinRM is not listening on port 5986 and then it starts to Listen on the port normally. Adding ignore errors didn't help as its "unreachable" error, and thus ansible just stop for all other kind of errors it can ignore errors.

Can you or anyone reading this thread try the play file? I just feel very stupid. 

P.s I will try the 5985 one, even though we have a policy not to allow http but I will try just for the sacke of testing.

**Comment by msghaleb** [comment]
Port 5985 is working fine! This is only happening for the SSL connection.

May be you can try it at ur end?

**Comment by msghaleb** [comment]
I tried today from MacOSX and also CentOS 6.5 same problem.

**Comment by nitzmahone** [comment]
We've struggled with this in our Windows CI env as well- looks to be related to the lowering of the default WinRM operation timeout to 20s in pywinrm 0.2.0 (guilty). For whatever reason, the first WinRM request to a newly created instance can take a _very_ long time, regardless of how long you wait to make it...

For now, you can work around this by increasing the winrm operation/read timeouts via inventory, eg:

```
ansible_winrm_operation_timeout_sec=60
ansible_winrm_read_timeout_sec=70
```

I may go back to pywinrm and increase the default timeouts again- the reason I set them much lower was to get faster failures on dead/unresponsive hosts. With the timeouts set this way, you have to wait 70s to find out a host is dead, where with the defaults, it's only 30s.

**Comment by msghaleb** [comment]
@nitzmahone Matt, you are very right! this has fixed my issue! I will work around it this way then.

Re: "I may go back to pywinrm and increase the default timeouts again" --> that could be a good idea, as I'm sure others are facing this too.

Many thanks 👍

**Comment by jhawkesworth** [comment]
Hey, great that there's a workaround for this.
I think I remember one of the features of Powershell 5.0 was 'reduced start up time' so I wonder if windows might start responding more quickly once Server 2016 is available.  I guess starting powershell is only part of the problem though, all sorts of dll caching and who knows what happening when the host first boots.  Oh and the possibility of daily scheduled Automatic Maintenance kicking off too.

**Comment by nitzmahone** [comment]
I've filed on bug for this on the pywinrm side just for tracking (https://github.com/diyan/pywinrm/issues/123)- what I'll probably do is implement polling on at least the open_shell operation in pywinrm (similar to what we already do for receive). That way, as long as the server keeps talking to us and saying "didn't finish but still working on it" we'll wait. That allows the steady state server operation timeouts and read timeouts to be kept low (leaving the aforementioned benefits of faster dead host detection) while still fixing this problem without user intervention.

**Comment by jhawkesworth** [comment]
That sounds like a great plan.

**Comment by msghaleb** [comment]
ok, I'll be closing this one here, as its part of the [pywinrm#123 ](https://github.com/diyan/pywinrm/issues/123)

**Comment by marcio-santos-zocdoc** [comment]
@nitzmahone thanks for the workaround. Can we add this workaround to the running ansible on windows documentation?

**Comment by dagwieers** [comment]
@marcio-santos-zocdoc I just implemented your suggestion, see: #24922

But don't wait for someone else to do it, anyone can update the documentation. You don't even need any special rights or know how to use git, you can edit documentation files directly from the Github web-interface: https://github.com/ansible/ansible/edit/devel/docs/docsite/rst/intro_windows.rst
